### ISSUE TITLE
Error using mkdocs serve and mathjax library installed locally

### ISSUE THREAD

**Issue by 1234qwerty1** [mkdocs/mkdocs#1351]
Hi, I think mkdocs is great but I'm having an issue with the mathjax library when trying to run the `mkdocs serve` command.  Below are what I have done and the error that I get. 

I have downloaded the mathjax library and unpacked it in my theme directory and edited main.html with the following:

```
 {% extends "base.html" %}
 {% block libs %}
      {{ super() }}
      <script src="{{ base_url }}/assets/javascripts/MathJax-master/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
  {% endblock %}
```
I have also added pymdown.arithmatex in my mkdocs.yaml file.

When I try to access my page after using the `mkdocs serve` command I get the following error.

```
[E 171115 21:15:22 web:1590] Uncaught exception GET /assets/javascripts/MathJax-master/extensions/MathMenu.js?V=2.7.2 (127.0.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8000', method='GET', uri='/assets/javascripts/MathJax-master/extensions/MathMenu.js?V=2.7.2', version='HTTP/1.1', remote_ip='127.0.0.1', headers={'Host': 'localhost:8000', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:56.0) Gecko/20100101 Firefox/56.0', 'Accept': '*/*', 'Accept-Language': 'en-GB,en;q=0.5', 'Accept-Encoding': 'gzip, deflate', 'Referer': 'http://localhost:8000/', 'Dnt': '1', 'Connection': 'keep-alive'})
    Traceback (most recent call last):
      File "/usr/lib64/python3.6/site-packages/tornado/web.py", line 1511, in _execute
        result = yield result
      File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 1055, in run
        value = future.result()
      File "/usr/lib64/python3.6/site-packages/tornado/concurrent.py", line 238, in result
        raise_exc_info(self._exc_info)
      File "<string>", line 4, in raise_exc_info
      File "/usr/lib64/python3.6/site-packages/tornado/gen.py", line 307, in wrapper
        yielded = next(result)
      File "/usr/lib64/python3.6/site-packages/tornado/web.py", line 2421, in get
        yield self.flush()
      File "/usr/lib64/python3.6/site-packages/tornado/web.py", line 947, in flush
        start_line, self._headers, chunk, callback=callback)
      File "/usr/lib64/python3.6/site-packages/tornado/http1connection.py", line 399, in write_headers
        data += self._format_chunk(chunk)
      File "/usr/lib64/python3.6/site-packages/tornado/http1connection.py", line 411, in _format_chunk
        "Tried to write more data than Content-Length")
    tornado.httputil.HTTPOutputError: Tried to write more data than Content-Length
[E 171115 21:15:22 web:1015] Cannot send error response after headers written
```

I have tried to narrow down the error and have found that the page loads successfully if I use `mkdocs build` and then open the generated page in a web browser. I have also tried adding the CDN to the `extra_javascript` option in my config file. This is also successful but I cannot use it all the time as I have a requirement to be offline and edit the page. I have also tried using multiple themes with the same results. I am on mkdocs v0.17.1.

So I am only seeing the error when trying to serve the page locally. I can't find a way around this after reading the documentation and the issues on github. Please can you help me resolve this?

**Comment by waylan** [comment]
At first I thought this was going to be another duplicate of #1336, which was fixed in version 0.17.2 released earlier today. After reading the report it appears to not be related, but could you upgrade to make sure?

Looking at the error, it appears that the server is raising an exception when it attempts to return the mathjax lib to the browser with a Content-Length related issue. My first guess would be that that is an upstream issue with the server, but I'm not sure. What happens if you use the `--no-livereload` option (which uses a simpler server config)?

I don't use mathjax personally, so I don't have much experience with this issue. Although I seem to recall seeing something like this somewhere before. @facelessuser do you have any insight?

**Comment by facelessuser** [comment]
Hmm.  Not sure.  They state the CDN is working through `extra_javascript`.  In my uses, I personally append the script CDN at the bottom of the page I want to use MathJax, so again a use case with the CDN. Unfortunately, I have not yet tried to load MathJax locally with MkDocs. I'd have to give it a try first to get a better understanding of what is required to set it up.  I'd have to also see how MkDocs handles that library.  Does it move all the related files, or just the main one you specify on preview?

Anyways, it would require more investigation.

**Comment by waylan** [comment]
I haven't tried it myself, but to host the lib locally you would need to include the lib file(s) in your `docs_dir`. On build, MkDocs copies everything which is not a `.md` file to the `site_dir` as-is (with binary read/write). The only recent changes we've made to that are in #1294, which I don't think should have any effect in this (I don't see how file permissions would effect content-length).

I wonder if its a file encoding issue. Reading with the wrong encoding could manifest itself as a file size being different than expected. @1234qwerty1 can you confirm that your files are encoded as UTF-8?

**Comment by facelessuser** [comment]
I'm not sure what the issue is, but MathJax is too big to be live previewing with its entire source.  Its a shame you can't tell MkDocs to copy a lib, but don't watch it.  It was so painfully slow trying to test this.

**Comment by waylan** [comment]
Sounds like the `--no-livereload` option is needed then.

**Comment by 1234qwerty1** [comment]
Thanks for your help. It is working with --no-livereload. Ideally it would work with the live reload but for now I'll find a workaround. I'm thinking I could host the library in a docker container and call that in my `extra_javascript`.

**Comment by waylan** [comment]
So it seems that the large volume of files is causing the "file watching" to break. I suppose if there was some way to exclude certain files/dirs form being watched, that would resolve the problem. Not sure how that would look though.

**Comment by 1234qwerty1** [comment]
What do you mean by "file watching"? If there were a way to (optionally) stop watching some files, I think the ones from the theme directory are best candidate. Then you could move  are all the files which are likely to be static to a theme and keep all files you want to edit in the docs directory or another theme which is watched if you are developing a theme. I'm not sure if you would be able to do that though.

**Comment by waylan** [comment]
> If there were a way to (optionally) stop watching some files, I think the ones from the theme directory are best candidate. 

With the [on_serve][1] event of the Plugin API you can alter the server instance. The server instance is an instance of [python-livereload][2], so you could use a plugin to alter or replace `server.watcher` to better meet your needs.

Regarding the theme directories, having them included in the watcher is necessary when developing a theme. Of course, its not necessary when editing content. Which makes more sense as the default is debatable, but given the low number of complaints with the current default, I'm inclined to leave it as-is. In fact, it is even recommended that the MathJax lib not be included in the theme, in which case, not "watching" the theme would make no difference.

Given the above, I'm closing this as wont-fix.

[1]: http://www.mkdocs.org/user-guide/plugins/#on_serve
[2]: https://github.com/lepture/python-livereload

**Comment by pradyunsg** [comment]
FYI: this is a bug in python-livereload -- see https://github.com/GaretJax/sphinx-autobuild/issues/71#issuecomment-681854580 for context.

**Comment by waylan** [comment]
Thanks for the report @pradyunsg. It is weird that Python-Livereload is searching for `<head>` in non-HTML files. That is definitely an upstream bug. The more curious thing to me is that this bug seems to only exhibit itself on Linux (which explains the reports about docker containers). Why doesn't this happen on Windows or Mac? In any event, that is an upstream issue. We should not need to fix it here. However, assuming a fix does happen, we need to update the version to the newest with the fix. Therefore, I am reopening this issue.

**Comment by pradyunsg** [comment]
Ahoy! I got push access to python-livereload. PRs to fix lepture/python-livereload#174 are welcome. They can definitely be based on https://github.com/executablebooks/sphinx-autobuild/pull/91! :)

**Comment by GrammAcc** [comment]
@waylan Sorry about the necrobump. I was going through old issues to see if there was anything suitable for a first contribution, and it looks like MkDocs no longer uses python-livereload. Should this issue be closed?
